Userland/Libraries/LibGUI/VimEditingEngine.cpp

```diff
@@ -1403,4 +1404,188 @@ void VimEditingEngine::move_to_next_empty_lines_block()
+        return;
+
+    m_commands.append(VimCommand("v", "N",
+        [&] {
+            switch_to_visual_mode();
+        }));
+    m_commands.append(VimCommand("<Escape>", "IV",
+        [&] {
+            switch_to_normal_mode();
+        }));
+    m_commands.append(VimCommand("i", "N",
+        [&] {
+            switch_to_insert_mode();
+        }));
+    m_commands.append(VimCommand("dd", "NV",
+        [&] {
+            if (m_motion.amount()) {
+                auto range = m_motion.get_repeat_range(*this, VimMotion::Unit::Line);
+                VERIFY(range.has_value());
+                yank(*range, Line);
+                m_editor->delete_text_range(*range);
+            } else {
+                yank(Line);
+                delete_line();
+            }
+            m_motion.reset();
+            m_previous_key = {};
+        }));
+}
+
+void VimEditingEngine::add_key(const KeyEvent& event)
+{
+    m_keys.append(VimKey(event.key(), event.text(), event.shift(), event.ctrl()));
+
+    VimMatchType match_type = VimMatchType::None;
+    for (const auto& command : m_commands) {
+        VimMatchType current_match = command.check_match(m_keys, m_vim_mode);
+        if (current_match == VimMatchType::Complete) {
+            match_type = VimMatchType::Complete;
+            command.execute();
+        } else if (current_match == VimMatchType::Partial) {
+            match_type = VimMatchType::Partial;
+        }
+    }
+    if (match_type == VimMatchType::None || match_type == VimMatchType::Complete) {
+        m_keys.clear();
+    }
+}
+
+bool VimEditingEngine::key_requires_shift(u32 code_point, KeyCode key)
+{
+    bool is_alphanumeric = is_ascii_alphanumeric(code_point);
+    bool is_alpha = is_ascii_alpha(code_point);
+
+    // Is a number
+    if (is_alphanumeric && !is_alpha)
+        return false;
+
+    if (is_alpha)
+        return is_ascii_upper_alpha(code_point);
+
+    return special_character_requires_shift(key);
+}
+
+bool VimEditingEngine::special_character_requires_shift(KeyCode key)
+{
+    static Vector<KeyCode> shift_required_keys = {
+        KeyCode::Key_ExclamationPoint,
+        KeyCode::Key_AtSign,
+        KeyCode::Key_Hashtag,
+        KeyCode::Key_Dollar,
+        KeyCode::Key_Percent,
+        KeyCode::Key_Circumflex,
+        KeyCode::Key_Ampersand,
+        KeyCode::Key_Asterisk,
+        KeyCode::Key_LeftParen,
+        KeyCode::Key_RightParen,
+        KeyCode::Key_Underscore,
+        KeyCode::Key_Plus,
+        KeyCode::Key_Pipe,
+        KeyCode::Key_Tilde,
+        KeyCode::Key_LeftBrace,
+        KeyCode::Key_RightBrace,
+        KeyCode::Key_Colon,
+        KeyCode::Key_DoubleQuote,
+        KeyCode::Key_LessThan,
+        KeyCode::Key_GreaterThan,
+        KeyCode::Key_QuestionMark
+    };
+    return shift_required_keys.contains_slow(key);
+}
+
+String VimKey::to_string() const
+{
+    return String::formatted("key_code: {} string:{} shift:{} ctrl:{}", (int)m_key_code, m_string, m_shift, m_ctrl);
```

Comment:
Please move debugging-related string creation to your `Formatter` specialization.

Additionally, please avoid C-style casts.
```suggestion
    return String::formatted("key_code: {} string:{} shift:{} ctrl:{}", static_cast<int>(m_key_code), m_string, m_shift, m_ctrl);